src/pudl/transform/ferc1.py

```diff
@@ -2801,6 +2807,73 @@ def process_instant_xbrl(self, df: pd.DataFrame) -> pd.DataFrame:
         return df.reset_index()
 
 
+class IncomeStatementFerc1TableTransformer(Ferc1AbstractTableTransformer):
+    """Transformer class for the :ref:`income_statement_ferc1` table."""
+
+    table_id: TableIdFerc1 = TableIdFerc1.INCOME_STATEMENT_FERC1
+    has_unique_record_ids: bool = False
+
+    def assign_record_id(
+        self, df: pd.DataFrame, source_ferc1: SourceFerc1
+    ) -> pd.DataFrame:
+        """Add a column identifying the original source record for each row.
+
+        Generate the ``record_id`` column in similar fashion to
+        :meth:`Ferc1AbstractTableTransformer.assign_record_id`, but use the table name
+        from the ``dbf_to_xbrl.csv`` during the :meth:`align_row_numbers_dbf` for the
+        DBF side of the table because this DBF input came from two seperate tables.
+
+        Args:
+            df: table to assign ``record_id`` to
+            table_name: name of table
+            source_ferc1: data source of raw ferc1 database.
+
+        Raises:
+            ValueError: If any of the primary key columns are missing from the DataFrame
+                being processed.
+            ValueError: If there are any null values in the primary key columns.
+            ValueError: If the resulting `record_id` column is non-unique.
+        """
+        logger.debug(
+            f"{self.table_id.value}: Assigning {source_ferc1.value} source record IDs."
+        )
+        pk_cols = self.renamed_table_primary_key(source_ferc1)
+        missing_pk_cols = set(pk_cols).difference(df.columns)
+        if missing_pk_cols:
+            raise ValueError(
+                f"{self.table_id.value} ({source_ferc1.value}): Missing primary key "
+                "columns in dataframe while assigning source record_id: "
+                f"{missing_pk_cols}"
+            )
+        if df[pk_cols].isnull().any(axis=None):
+            raise ValueError(
+                f"{self.table_id.value} ({source_ferc1.value}): Found null primary key "
+                "values.\n"
+                f"{df[pk_cols].isnull().any()}"
+            )
+        # assign the source_table_id column based on the source
+        if source_ferc1 == SourceFerc1.DBF:
+            # sched_table_name is a table name. was added in align_row_numbers_dbf
+            df = df.assign(source_table_id=lambda x: x.sched_table_name)
```

Comment:
see my pr comment about this. i don't think we can do this across the board w/o a few extra flourishes that i think would make this more complex overall.

Reply:
if we wanted to do this always... I would suggest we always employ `align_row_numbers_dbf` by default and convert the param to a boolean param instead of a table name (the table name(s) can be looked up from the extract dict).

Reply:
* The extract step does sound like the better place to get `sched_table_name` from.
* Maybe just a special `get_source_table_id()`?

Reply:
okay i've converted the `source_table_id` to be the special case for this table so this larger method is not mostly duplication. This does mean that `source_table_id` now takes a df... which is almost never used!